client/rest/src/test/java/org/elasticsearch/client/DeadHostStateTests.java

```diff
@@ -0,0 +1,120 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client;
+
+import java.util.concurrent.TimeUnit;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.lessThan;
+
+public class DeadHostStateTests extends RestClientTestCase {
+
+    private static long[] EXPECTED_TIMEOUTS_SECONDS = new long[]{60, 84, 120, 169, 240, 339, 480, 678, 960, 1357, 1800};
+
+    public void testInitialDeadHostState() {
+        DeadHostState deadHostState = new DeadHostState();
+        assertThat(deadHostState.getDeadUntilNanos(), greaterThan(System.nanoTime()));
+        assertThat(deadHostState.getFailedAttempts(), equalTo(1));
+    }
+
+    public void testDeadHostStateFromPrevious() {
+        DeadHostState previous = new DeadHostState();
+        int iters = randomIntBetween(5, 30);
+        for (int i = 0; i < iters; i++) {
+            DeadHostState deadHostState = new DeadHostState(previous);
+            assertThat(deadHostState.getDeadUntilNanos(), greaterThan(previous.getDeadUntilNanos()));
+            assertThat(deadHostState.getFailedAttempts(), equalTo(previous.getFailedAttempts() + 1));
+            previous = deadHostState;
+        }
+    }
+
+    public void testShallBeRetried() {
+        ConfigurableTimeSupplier timeSupplier = new ConfigurableTimeSupplier();
+        DeadHostState deadHostState = null;
+        for (int i = 0; i < EXPECTED_TIMEOUTS_SECONDS.length; i++) {
+            long expectedTimeoutSecond = EXPECTED_TIMEOUTS_SECONDS[i];
+            timeSupplier.nanoTime = 0;
+            if (i == 0) {
+                deadHostState = new DeadHostState(timeSupplier);
+            } else {
+                deadHostState = new DeadHostState(deadHostState, timeSupplier);
+            }
+            for (int j = 0; j < expectedTimeoutSecond; j++) {
+                timeSupplier.nanoTime += TimeUnit.SECONDS.toNanos(1);
+                assertThat(deadHostState.shallBeRetried(), is(false));
+            }
+            int iters = randomIntBetween(5, 30);
+            for (int j = 0; j < iters; j++) {
+                timeSupplier.nanoTime += TimeUnit.SECONDS.toNanos(1);
+                assertThat(deadHostState.shallBeRetried(), is(true));
+            }
+        }
+    }
+
+    public void testDeadHostStateTimeouts() {
```

Comment:
It *feels* to me like this should be part of `testDeadHostStateFromPrevious`. If you really want to have a version of the test that use `System.nanoTime` I'd name it with `DefaultTimeProvder` in the method name and name the method that uses the `ConfigurableTimeSupplier` more like `testDeadHostStateFromPrevious`.